src/filetree.rs

```diff
@@ -272,20 +280,41 @@ pub(crate) struct ApplyUpdateOptions {
 // Let's just fork off a helper process for now.
 #[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
 pub(crate) fn syncfs(d: &openat::Dir) -> Result<()> {
-    use rustix::fd::BorrowedFd;
     use rustix::fs::{Mode, OFlags};
     let d = unsafe { BorrowedFd::borrow_raw(d.as_raw_fd()) };
     let oflags = OFlags::RDONLY | OFlags::CLOEXEC | OFlags::DIRECTORY;
     let d = rustix::fs::openat(d, ".", oflags, Mode::empty())?;
     rustix::fs::syncfs(d).map_err(Into::into)
 }
 
+/// Copy from src to dst at root dir
 #[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
-fn tmpname_for_path<P: AsRef<Path>>(path: P) -> std::path::PathBuf {
-    let path = path.as_ref();
-    let mut buf = path.file_name().expect("filename").to_os_string();
-    buf.push(TMP_PREFIX);
-    path.with_file_name(buf)
+fn copy_dir(root: &openat::Dir, src: &str, dst: &str) -> Result<()> {
+    let rootfd = unsafe { BorrowedFd::borrow_raw(root.as_raw_fd()) };
+    let r = unsafe {
+        Command::new("cp")
+            .args(["-a"])
+            .arg(src)
+            .arg(dst)
+            .pre_exec(move || rustix::process::fchdir(rootfd).map_err(Into::into))
+            .status()?
+    };
+    if !r.success() {
+        anyhow::bail!("Failed to copy {src} to {dst}");
+    }
+    log::debug!("Copy {src} to {dst}");
+    Ok(())
+}
+
+/// Get first sub dir and tmp sub dir for the path
+/// "fedora/foo/bar" -> ("fedora", ".btmp.fedora")
+/// "foo" -> ("foo", ".btmp.foo")
+#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
+fn get_subdir(path: &Utf8Path) -> Result<(String, String)> {
+    let buf = path.iter().next().unwrap().to_owned();
+    let mut buf_tmp = buf.clone();
+    buf_tmp.insert_str(0, TMP_PREFIX);
+    Ok((buf, buf_tmp))
```

Comment:
I think this can look like this instead:

```
let first = path.iter().next().ok_or_else(|| anyhow!("Invalid path: {path}"))?;
let mut tmp = first.as_str().to_owned();
tmp.insert_str(0, TMP_PREFIX);
Ok((first, tmp))
```

The return value would then be `Result<(&Utf8Path, String)>` - notice we didn't need to make a copy of the first one, we can just return a borrowed value.

Reply:
LGTM, thanks, updated.